### DISCUSSION TITLE
Memory job parameter

### DISCUSSION THREAD

**Discussion by arrabito** [DIRACGrid/DIRAC#7509] [Q&A]
Dear all,

As far as I know the 2 memory related job parameters:  'Memory', 'MemoryUsed'
are not reliable or at least I don't understand their meaning.

Can you please confirm?

What it seems a reliable information instead is the RSS information logged in the pilot output.

Can you also confirm?

Thank you.

**Comment by fstagni** [comment]
(quickly, no time now). Why do you think they are not reliable?

MemoryUsed parameter is generated with 

```
resource.getrusage(resource.RUSAGE_SELF).ru_maxrss + resource.getrusage(resource.RUSAGE_CHILDREN).ru_maxrss
```

While what you see in the pilot is calculated with 

```
    def memoryUsage(self, withChildren=False):
        """
        Returns the memory usage of the process in MB
        """
        # Information is returned in bytes
        rss = self.process.memory_info().rss
        if withChildren:
            for child in self.process.children(recursive=True):
                rss += child.memory_info().rss
        # converted to MB
        return S_OK(rss / float(2**20))
```

**Reply by arrabito** [reply]
OK so probably I've just not understood the difference between the 2.
From your snapshot, I understand that what I see in the the pilot is the memory used by a single process and corresponds to what is used by the application (in case of single-core applications).

However which is the meaning of the MemoryUsed parameter? 

The values of the 2 parameters are very different, e.g.

`LoadAvg: 16 MemUsed: 140864.0 kb Job Vsize: 4323160.0 kb Job RSS: 2368944.0 kb DiskSpace: 4090583.0 MB CPU: 10:23:06 (h:m:s) WallClock: 37803.93 s`

and I can't immediately see the meaning of the MemoryUsed parameter.

Also is the unit MB for the 2 parameters? 

In the log is mentioned "kb" while from the code it seems to be in MB...

Thank you.

**Comment by fstagni** [comment]
Taking one random LHCb job:

In Jobs parameters:

```
MemoryUsed(kb) 30668
Memory(kB) 170880776
```

In pilot log:

```
==========================================================
Watchdog will report NodeInformation
==========================================================
Memory(kB) = 170880776 


...

==========================================================
Watchdog will report UsageSummary
==========================================================
...
MemoryUsed(kb) = 30668.0
...
```

I do not see discrepancies.

BTW:

```python
result["Memory(kB)"] = int(psutil.virtual_memory()[1] / 1024)
```
which is the available virtual memory.


Are you running single or multi-core applications?

**Comment by arrabito** [comment]
We are using single core applications, but we use :

`LocalCEType=Pool`

I don't know if it makes a difference in getting memory informations.

Anyway, I also don't see any discrepancies between Job Parameters and what is reported in the pilot log. 

However the MemoryUsed parameter does not correspond at all (by orders of magnitudes) to what we have measured for our application independentely.

While the RSS computed by:

```
    def memoryUsage(self, withChildren=False):
        """
        Returns the memory usage of the process in MB
        """
        # Information is returned in bytes
        rss = self.process.memory_info().rss
        if withChildren:
            for child in self.process.children(recursive=True):
                rss += child.memory_info().rss
        # converted to MB
        return S_OK(rss / float(2**20)
```
i.e. using psutil corresponds well to our measurements.

As far as I understand the memory computed with psutil corresponds to a given process (the one associated to the job I guess), while:

`resource.getrusage(resource.RUSAGE_SELF).ru_maxrss + resource.getrusage(resource.RUSAGE_CHILDREN).ru_maxrss`

gives the memory used by the caller, which in this case I guess is the JobWrapper. Is that correct?
In principle since child process are also considered this should give similar result as psutil with the difference that RSS is reported at different time intervals, while the MemoryUsed parameter is computed taking the difference between the initial value and the final one:

```
        # MemoryUsed
        if "MemoryUsed" in self.parameters:
            memory = self.parameters["MemoryUsed"]
            if memory:
                summary["MemoryUsed(kb)"] = abs(float(memory[-1]) - float(self.initialValues["MemoryUsed"]))
            else:
                summary["MemoryUsed(kb)"] = math.nan
```
which is fine.

So I'm wondering if:

`resource.getrusage(resource.RUSAGE_SELF).ru_maxrss + resource.getrusage(resource.RUSAGE_CHILDREN).ru_maxrss`

correctly takes into account the memory used by the process associated to the job or it takes something else.

To give you an example for a given job we have RSS of the order of:

RSS: 2887944.0 kb 

and

MemoryUsed (kb): 69912.0

Can you please explain the large difference between these 2 values?

**Comment by fstagni** [comment]
Few things:
- Using `Pool` as Local CE should not pose an issue
- a pilot can (will) run more than one job, and in the case of `Pool` they will run in parallel. This I am afraid may result in a rather messy pilot log. I will see if at least the printouts can be improved a bit
- There's some historical confusion about MB and KB in the printouts, I see if I can improve.
- I think I found a bug in how RSS is reported, I will re-check and correct
- When we talk about "MemoryUsed" we are effectively reporting RSS Usage, so I will remove one of the two.

**Reply by fstagni** [reply]
Please check https://github.com/DIRACGrid/DIRAC/pull/7518

**Comment by arrabito** [comment]
Thanks a lot for the PR.

It looks fine to me.

I just have a minor comment that I will put directly on the PR.

Then, about using Pool as Local CE, if NumeberOfProcessors is set = 1, then there will be no jobs in parallel right? So on in this case the log shouldn't be messy.

**Reply by fstagni** [reply]
> Then, about using Pool as Local CE, if NumeberOfProcessors is set = 1, then there will be no jobs in parallel right? So on in this case the log shouldn't be messy.

Correct.
